### PR TITLE
KAFKA-10003: Mark KStream.through() as deprecated

### PR DESCRIPTION
**Pull request by mjsax**
- part of KIP-221

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @lkokhreidze @vvcephei 

Also updates the Scala API...

File: docs/streams/developer-guide/dsl-api.html
Comment (mjsax) [inline docs/streams/developer-guide/dsl-api.html:3694]: The diff is weird because the part above repeats below. The actual deletes starts here.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1084]: Not sure if this update is necessary. This method is deprecated itself.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:849]: Not 100% sure if we should remove this now, or when we remove `through()`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:924]: Just simplifying this one.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/RepartitionedInternal.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/RepartitionedInternal.java:24]: Must be public to be visible in Scala

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:291]: Replicated the test for `through()` for `repartition()`.

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java:501]: We still need to test this, because topics using this pattern are still consider _intermediate_ topics and the `--intermediat-topic` flag in `StreamsResetter` is still useful and not changed.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java:217]: Using `to()` and `steam()` is "simpler" as we cleanup topics in-between (and thus avoid internal topics).

We could of course also use `repartition()`.

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationWithSslTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationWithSslTest.java:45]: side cleanup

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:596]: replicating test

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:1415]: replicating test cases

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:1478]: replicating test

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:1511]: replicating test

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestClient.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestClient.java:136]: As above. Avoid internal topics.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala
Comment (mjsax) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:221]: There is no `KTable#through()` method.

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/ProducedTest.scala
Comment (mjsax) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/ProducedTest.scala:40]: Side cleanup (was originally copied from `ConsumedTest` but not updated correctly)

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Repartitioned.scala
Comment (mjsax) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Repartitioned.scala:35]: I just named all method `with` in alignment to the other Scala helper classes.

Also noticed, that all helper classed only have static methods... Is not by design? Seems we are missing something here? If there is more than one optional parameter, it seems we should have non-static method to allow method chaining? (Could be fixed in a follow up PR)

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: It's a little nice for future reference when we also say when it became deprecated, such as "since 2.6".

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:849]: I'd agree with removing it. I guess if you want to preserve it in some fashion, you could add the opposite statement to the `through()` documentation.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Repartitioned.scala
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Repartitioned.scala:35]: We'd have to change from `object` to `class` or `case class` (which would have been my preference to begin with), since `object`s can only have static members.

Probably, this ship has sailed for now, and we should just keep doing what the other similar classes are doing. Since we've found so much wackiness in the Scala API since it was introduced, it might be a good idea to consider revamping the whole thing from scratch some day.

**Review by vvcephei — COMMENTED** [review]
Just found another set of zombie comments I meant to send some time in the past. I'll continue my review.

File: docs/streams/developer-guide/dsl-api.html
Comment (vvcephei) [inline docs/streams/developer-guide/dsl-api.html]:
```suggestion
                            It is recommended to repartition the topic with fewer partitions to match the larger partition number of avoid bottlenecks.
```

File: docs/streams/upgrade-guide.html
Comment (vvcephei) [inline docs/streams/upgrade-guide.html]:
```suggestion
        If you need to write into and read back from a topic that you mange, you can fall back to use <code>KStream.to()</code> in combination with <code>StreamsBuilder#stream()</code>.
```

File: docs/streams/upgrade-guide.html
Comment (vvcephei) [inline docs/streams/upgrade-guide.html]:
```suggestion
        Please refer to the <a href="/{{version}}/documentation/streams/developer-guide/dsl-api.html">developer guide</a> for more details about <code>KStream.repartition()</code>.
```

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1084]: Might as well make this update, since we may remove the methods at different times.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/RepartitionedInternal.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/RepartitionedInternal.java:24]: It's worth noting that it only needs to be visible for the scala _tests_ that verify the scala Repartitioned builder results in a correctly configured object. For the public API, we only convert a scala Repartitioned to a java Repartitioned.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: My opinion is that it's generally better not to suppress but instead just deprecate this method as well. It's not really that important for tests, since no one else is going to call the method, so feel free to take or leave the advice.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:291]: Thanks!

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java:501]: I'm wondering if we should continue testing with `through`, to ensure it continues to work. WDYT?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java]: This would be a case where I would advocate more strongly to deprecate _this_ method, to avoid accidentally "hiding" the deprecation from callers.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:221]: Oops...

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala]: I think we'd prefer:
```suggestion
  def repartition(implicit repartitioned: Repartitioned[K, V]): KStream[K, V] =
```
similar to groupByKey, although I'm admittedly not sure if it actually matters.

**Review by vvcephei — COMMENTED** [review]
Thanks, @mjsax ! Completed my full pass.

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java:501]: Well, `through()` is literally implemented as `to()` + `stream()`... But I can revert and add a suppress annotation, too.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: Not sure why? If I use 2.6 why do I can if it was deprecated in 2.4 or 2.2 or 2.6? It's deprecated in the version I use now. Why would I care about older versions?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java]: Well, but then we need to add more suppression or deprecation upstream. Does not seem worth for testing code

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java:501]: On re-reading, I realize I misunderstood the situation. I revert my comment :grimacing: .

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: For one thing, it's nice for _us_, so we can easily tell when it's been deprecated "long enough" to remove. I can recall trudging through git history in the past to figure this out.

For users, maybe you don't care, but I personally find it nice when my libraries do this for me. It's just good bookkeeping, and it gives me some confidence that the maintainers are doing proper, tidy maintenance.

If it provides a "third party" supporting opinion, the Scala language designers thought this was important enough to build it in as a separate field of the "deprecated" annotation: https://docs.scala-lang.org/tour/annotations.html

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java]: This is exactly the point!

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala]: We should provide the arguments to the annotation to explain the deprecation. This gives better IDE integration than the ScalaDoc (which we won't need anymore):

```suggestion
  @deprecated("use `repartition()` instead", "2.6.0")
```

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:279]: I just noticed that we have no test for this operator (or for through). Should we add one?

**Review by vvcephei — COMMENTED** [review]
Thanks for the update, @mjsax ! I replied above to the threads, and had just two more new comments.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala
Comment (mjsax) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:279]: Not sure what we can/should test?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java]: Updated

**Review by lkokhreidze — APPROVED** [review]
Thanks for picking this up @mjsax , lgtm.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:279]: We have previously had embarrassing bugs like, "It's not possible to write code that compiles using the Scala DSL". If we had had even the most trivial test written for those DSL methods, we would never have released those bugs. So, I'd just recommend creating _any_ topology that contains this operator and maybe using TTD to pipe a single record through it to ensure that it doesn't throw any runtime exceptions when you use it.

**Review by vvcephei — APPROVED** [review]
Thanks for this thorough (and thoroughly awesome) PR, @mjsax !

I responded to the question about the scala API test, above; otherwise I'm +1.

**Comment by mjsax** [comment]
Added the test. Will merge after Jenkins passed.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: Could we use ```{@link #repartition(Repartitioned)}``` ?

**Review by vvcephei — APPROVED** [review]
Thanks so much, @mjsax !

**Comment by mjsax** [comment]
Java 8 passed.
Java 11:
```
org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[false]
```
Java 14:
```
org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[false]
```

**Comment by mjsax** [comment]
Java 8 and Java 11 passed.
Java 14:
```
org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[true]
```

**Pull request merged**
